src/argilla/monitoring/asgi.py

```diff
@@ -113,30 +150,60 @@ def __init__(
         **kwargs
     ):
         super().__init__(*args, **kwargs)
+
         self._endpoint = api_endpoint
         self._dataset = dataset
         self._records_mapper = records_mapper or text_classification_mapper
+        self._queue = None
+        self._worker_task = None
+        self._initialized = False
+
+    def init(self):
+        if self._initialized:
+            return
+
         self._queue = Queue()
         self._worker_task = threading.Thread(
-            target=self.__worker__, name=argillaLogHTTPMiddleware.__name__, daemon=True
+            target=self.__worker__,
+            name=ArgillaLogHTTPMiddleware.__name__,
+            daemon=True,
         )
         self._worker_task.start()
+        atexit.register(self._worker_task.join)
+        self._initialized = True
 
     async def dispatch(
-        self, request: Request, call_next: RequestResponseEndpoint
+        self,
+        request: Request,
+        call_next: RequestResponseEndpoint,
     ) -> Response:
+        if not self._initialized:
+            self.init()
+
         if self._endpoint != request.url.path:  # Filtering endpoint path
             return await call_next(request)
 
         content_type = request.headers.get("Content-type", None)
-        if "application/json" not in content_type:
-            return await call_next(request)
+
+        if content_type is not None:
+            if "application/json" not in content_type:
+                return await call_next(request)
 
         cached_request = CachedJsonRequest(
-            scope=request.scope, receive=request.receive, send=request._send
+            scope=request.scope,
+            receive=request.receive,
+            send=request._send,
         )
-        # Must read body before call_next
-        inputs = await cached_request.json()
+
+        # Must read body or query params before call_next
+        if cached_request.method == "POST":
```

Comment:
The `PUT` method is still missing, isn't it?

In that case, I think we could do something like this:

```suggestion
        if cached_request.method in ["POST", "PUT"]:
```